src/rpc/protocol.cpp

```diff
@@ -84,14 +91,20 @@ bool GenerateAuthCookie(std::string *cookie_out)
      * these are set to 077 in init.cpp unless overridden with -sysperms.
      */
     std::ofstream file;
-    fs::path filepath = GetAuthCookieFile();
-    file.open(filepath.string().c_str());
+    fs::path filepath_tmp = GetAuthCookieFile(true);
+    file.open(filepath_tmp.string().c_str());
     if (!file.is_open()) {
-        LogPrintf("Unable to open cookie authentication file %s for writing\n", filepath.string());
+        LogPrintf("Unable to open cookie authentication file %s for writing\n", filepath_tmp.string());
         return false;
     }
     file << cookie;
     file.close();
```

Comment:
I've not tested but [here](http://www.microhowto.info/howto/atomically_rewrite_the_content_of_a_file.html) says:
> A common misconception is that calling close will necessarily perform an implicit fsync.

From the [docs](http://www.cplusplus.com/reference/fstream/ofstream/close/) there is no explicit mention to flush, only:

> Any pending output sequence is written to the file.

Reply:
In other words, call `file.flush()` before close?

Reply:
You're having another misconception there: fsync is not necessary for other processes to see the result! Only to make sure that the data hits disk, in case of crashes etc. Which is not relevant here.

Reply:
Let's ask @theuni just in case.

Reply:
I think the only concern is that libc may buffer the write as long as the file is open, in which case a flush would be useful to make it visible to the OS (and thus other processes). However, as you're closing the file anyway, I don't believe that matters.